### PR TITLE
Add additional features to the comport list

### LINKED ISSUE #16: API: Rework Comports Command
The current comports command implementation on the API simply returns a list of serial ports. We should consider expanding this to include more information, such as:

- The associated device (e.g. CP2102 USB to UART bridge)
- Whether that comport has an open connection tied to it

This issue will change the client-server contract and will require all clients to update their implementation. However, it also opens up some great possibilities w.r.t. dashboard, since we could display device names and use the connection status to set the color of the dot for each device in the list.

cc: @favillat

### PR DESCRIPTION
**Pull request by ETSells**
#### Issue: N/A 

*An issue does not yet exist for tracking on the dashboard. However, as the API side changes the contract of the /comports endpoint, I wanted to make sure I provided a reference implementation for my own testing and so that I don't force you guys to update.*

## Description
Provides a reference implementation for the changes implemented to resolve https://github.com/SunDevilRocketry/SDEC-API/issues/16.

Child of: https://github.com/SunDevilRocketry/SDEC-API/pull/18

There are two features implemented here. One: the /comports route now returns a json with {port name, device description}. The comport list now reflects this. Two, a route at /comports/active has been added. This returns a string with the currently active comport to make it easier for multiple users to connect to the same device. This also carries a comport list update, where we now set the red dot to yellow if a device is able to be connected to.

Screenshot:
image

## Originator Checklist
- [ ] Title matches the form "GHI #<number> - <description of code change> [<Target Branch>]"
- [X] Target branch is correct
- [ ] Unit Tests have been posted in issue <if applicable,can be super simple like a screenshot>
- [ ] Issue has been linked to this PR
- [ ] Changes generate no new warnings
    - merge main into your branch to resolve conflicts before opening PR!

<br>

---

I like the new PR template! I'm a bit confused about the title field though, and I also think you might be misusing the term "unit tests"

### CODE REVIEW

**Comment by favillat** [comment]
I like the red-yellow status, though I'm curious whether it might provide a cleaner experience to only show connectable COM ports and hide the unavailable ones altogether, Unless there is a use for showing the red ones?

**Comment by ETSells** [comment]
It's not necessarily "connectable" vs not. The yellow status lights indicate that there is an *existing* connection i.e. another client has connected the API to that port. I unfortunately don't really have any way of determining whats connectable or not unless we make a giant lookup table of every device name for the CP210x or STM CDC.

File: src/components/widgets/BoardStatusWidget.js [hunk 1]
```diff
@@ -123,7 +133,16 @@ export const BoardStatusWidget = ({
                 (
                 <div className="flex flex-col w-full">
                 <MockBoard onMockConnected={onMockConnected} />
-                {boards.map((port, i) => ( <COMBoard key={i} name={port} onConnect={onConnect} index={i} />))}
+                {boards.map(({ port, device_description }, i) => (
```
Comment (favillat) [inline src/components/widgets/BoardStatusWidget.js:141]: nitpick, naming: follow camelCase formatting

**Comment by ETSells** [comment]
Actually, you might have a point -- we should consider dropping the other comports if comports/active matches any of the ports from the list (except for the mock, that should stay imho)

File: src/components/widgets/BoardStatusWidget.js (hunk 1 above)
Comment (ETSells) [reply src/components/widgets/BoardStatusWidget.js:141]: willfix to camelCase. My bad, firmware + sdec use lower snake_case, its a habit atp.

Comment (favillat) [reply src/components/widgets/BoardStatusWidget.js:141]: thanks! everything else looks solid, I like the descriptions a lot. I'll go in and write one for the mock connection (a bit unnecessary, mainly to keep consistency)

Comment (ETSells) [reply src/components/widgets/BoardStatusWidget.js:141]: Ah thanks for handling the fix for me! I tried to look up the accent styles and couldn't find them w/ find all, so glad thats more idiomatic to the project. Altogether really cool!!

**Comment by favillat** [comment]
This looks ready to me! unless you want to add hiding comports ill go ahead and approve; side note on the template, I literally just ripped the one we use at work lol, I'm planning on updating it to better suit SDR/new members

**Comment by ETSells** [comment]
I probably won't get to it for a while, so I suppose we'll see. I try to minimize how much I do on the dashboard bc this isnt a strength of mine. I can probably do it on the API side when I have time again.

Thanks for taking a look! Remember that this is in response to the API PR which modifies the contract, so this won't work without the API side being merged and it'll all need to go in at once.

**Comment by favillat** [comment]
yeah, I can go ahead and do it it's no pressure ik you have stuff to worry about on firmware. so I'll wait to merge until[ api comport rework](https://github.com/SunDevilRocketry/SDEC-API/issues/16) is merged?

**Comment by ETSells** [comment]
yep! also, if possible, since leo and ethan aren't super comfortable running their end of the project I have you tagged as a reviewer on the parent and all three children. i don't know your comfort level with python but imho it should be pretty readable? worst case I had coderabbit review stuff and we don't have any real SMEs so i can just send it 💀

**Comment by favillat** [comment]
yea, I mean I'm pretty average with python but I'd be happy to review anything you send my way

**Comment by ETSells** [comment]
I believe you're tagged on all of them. I've consolidated some of the PRs that do similar things in this area. I've added the requested port filtering on the API side:

image


Also, I've done a bit more testing of edge cases and improper use, and most of my API work looks like it's made the system way more robust. We do a significantly better job of handling errors server-side provided the client is equipped to handle 204 no content statuses + 400/500 errors.

Also also, I recommend increasing the rate at which the /comports route is called to ~once a second prior to establishing a connection to ensure that the available boards status updates.
